library/CMakeLists.txt

```diff
@@ -10,6 +10,14 @@ if(NOT DEFINED MBEDTLS_DIR)
     set(MBEDTLS_DIR ${CMAKE_SOURCE_DIR})
 endif()
 
+# If set, make MBEDTLS_CONFIG_FILE and MBEDTLS_USER_CONFIG_FILE into PATHs
```

Comment:
Not working for me in what I thought would be a typical use case.

```
mkdir build-cmake-thread
cd build-cmake-thread
cmake -DMBEDTLS_CONFIG_FILE=../configs/config-thread.h ..
make
```
The build of the library seems to go fine, but then the first program fails.
```
…
[ 20%] Linking C static library libmbedtls.a
[ 20%] Built target mbedtls
Scanning dependencies of target crypt_and_hash
[ 20%] Building C object programs/aes/CMakeFiles/crypt_and_hash.dir/crypt_and_hash.c.o
[ 20%] Linking C executable crypt_and_hash
/usr/bin/ld: ../../CMakeFiles/mbedtls_test.dir/tests/src/helpers.c.o: in function `mbedtls_test_platform_setup':
helpers.c:(.text+0x1b): undefined reference to `mbedtls_platform_setup'
/usr/bin/ld: ../../CMakeFiles/mbedtls_test.dir/tests/src/helpers.c.o: in function `mbedtls_test_platform_teardown':
helpers.c:(.text+0x37): undefined reference to `mbedtls_platform_teardown'
/usr/bin/ld: ../../CMakeFiles/mbedtls_test.dir/tests/src/psa_crypto_helpers.c.o: in function `mbedtls_test_psa_purge_key_storage':
psa_crypto_helpers.c:(.text+0xf9): undefined reference to `psa_destroy_persistent_key'
/usr/bin/ld: ../../CMakeFiles/mbedtls_test.dir/tests/src/psa_crypto_helpers.c.o: in function `mbedtls_test_psa_purge_key_cache':
psa_crypto_helpers.c:(.text+0x14c): undefined reference to `psa_purge_key'
/usr/bin/ld: ../../CMakeFiles/mbedtls_test.dir/tests/src/psa_crypto_helpers.c.o: in function `mbedtls_test_helper_is_psa_leaking':
psa_crypto_helpers.c:(.text+0x189): undefined reference to `mbedtls_psa_get_stats'
/usr/bin/ld: ../../CMakeFiles/mbedtls_test.dir/tests/src/psa_exercise_key.c.o: in function `check_key_attributes_sanity':
psa_exercise_key.c:(.text+0xaa): undefined reference to `psa_get_key_attributes'
/usr/bin/ld: psa_exercise_key.c:(.text+0x2f2): undefined reference to `psa_reset_key_attributes'
/usr/bin/ld: ../../CMakeFiles/mbedtls_test.dir/tests/src/psa_exercise_key.c.o: in function `exercise_mac_key':
psa_exercise_key.c:(.text+0x3ee): undefined reference to `psa_mac_sign_setup'
/usr/bin/ld: psa_exercise_key.c:(.text+0x434): undefined reference to `psa_mac_update'
/usr/bin/ld: psa_exercise_key.c:(.text+0x481): undefined reference to `psa_mac_sign_finish'
/usr/bin/ld: psa_exercise_key.c:(.text+0x4fd): undefined reference to `psa_mac_verify_setup'
/usr/bin/ld: psa_exercise_key.c:(.text+0x543): undefined reference to `psa_mac_update'
/usr/bin/ld: psa_exercise_key.c:(.text+0x590): undefined reference to `psa_mac_verify_finish'
/usr/bin/ld: psa_exercise_key.c:(.text+0x5da): undefined reference to `psa_mac_abort'
/usr/bin/ld: ../../CMakeFiles/mbedtls_test.dir/tests/src/psa_exercise_key.c.o: in function `exercise_cipher_key':
psa_exercise_key.c:(.text+0x70c): undefined reference to `psa_get_key_attributes'
/usr/bin/ld: psa_exercise_key.c:(.text+0x878): undefined reference to `psa_cipher_encrypt_setup'
/usr/bin/ld: psa_exercise_key.c:(.text+0x8d1): undefined reference to `psa_cipher_generate_iv'
/usr/bin/ld: psa_exercise_key.c:(.text+0x92b): undefined reference to `psa_cipher_update'
/usr/bin/ld: psa_exercise_key.c:(.text+0x993): undefined reference to `psa_cipher_finish'
/usr/bin/ld: psa_exercise_key.c:(.text+0xa3b): undefined reference to `psa_cipher_decrypt_setup'
/usr/bin/ld: psa_exercise_key.c:(.text+0xa8f): undefined reference to `psa_cipher_set_iv'
/usr/bin/ld: psa_exercise_key.c:(.text+0xae8): undefined reference to `psa_cipher_update'
/usr/bin/ld: psa_exercise_key.c:(.text+0xb50): undefined reference to `psa_cipher_finish'
/usr/bin/ld: psa_exercise_key.c:(.text+0xbea): undefined reference to `psa_cipher_abort'
/usr/bin/ld: psa_exercise_key.c:(.text+0xbf9): undefined reference to `psa_reset_key_attributes'
/usr/bin/ld: ../../CMakeFiles/mbedtls_test.dir/tests/src/psa_exercise_key.c.o: in function `exercise_aead_key':
psa_exercise_key.c:(.text+0xd3e): undefined reference to `psa_get_key_attributes'
/usr/bin/ld: psa_exercise_key.c:(.text+0xe6a): undefined reference to `psa_aead_encrypt'
/usr/bin/ld: psa_exercise_key.c:(.text+0xf22): undefined reference to `psa_aead_decrypt'
/usr/bin/ld: psa_exercise_key.c:(.text+0xf67): undefined reference to `psa_reset_key_attributes'
/usr/bin/ld: ../../CMakeFiles/mbedtls_test.dir/tests/src/psa_exercise_key.c.o: in function `exercise_signature_key':
psa_exercise_key.c:(.text+0x1399): undefined reference to `psa_sign_hash'
/usr/bin/ld: psa_exercise_key.c:(.text+0x143a): undefined reference to `psa_verify_hash'
/usr/bin/ld: psa_exercise_key.c:(.text+0x1677): undefined reference to `psa_sign_message'
/usr/bin/ld: psa_exercise_key.c:(.text+0x1718): undefined reference to `psa_verify_message'
/usr/bin/ld: ../../CMakeFiles/mbedtls_test.dir/tests/src/psa_exercise_key.c.o: in function `exercise_asymmetric_encryption_key':
psa_exercise_key.c:(.text+0x1ac1): undefined reference to `psa_asymmetric_encrypt'
/usr/bin/ld: psa_exercise_key.c:(.text+0x1b49): undefined reference to `psa_asymmetric_decrypt'
/usr/bin/ld: ../../CMakeFiles/mbedtls_test.dir/tests/src/psa_exercise_key.c.o: in function `mbedtls_test_psa_setup_key_derivation_wrap':
psa_exercise_key.c:(.text+0x1bf3): undefined reference to `psa_key_derivation_setup'
/usr/bin/ld: psa_exercise_key.c:(.text+0x1c47): undefined reference to `psa_key_derivation_input_bytes'
/usr/bin/ld: psa_exercise_key.c:(.text+0x1c86): undefined reference to `psa_key_derivation_input_key'
/usr/bin/ld: psa_exercise_key.c:(.text+0x1cca): undefined reference to `psa_key_derivation_input_bytes'
/usr/bin/ld: psa_exercise_key.c:(.text+0x1d2f): undefined reference to `psa_key_derivation_input_bytes'
/usr/bin/ld: psa_exercise_key.c:(.text+0x1d6e): undefined reference to `psa_key_derivation_input_key'
/usr/bin/ld: psa_exercise_key.c:(.text+0x1db2): undefined reference to `psa_key_derivation_input_bytes'
/usr/bin/ld: psa_exercise_key.c:(.text+0x1e0f): undefined reference to `psa_key_derivation_set_capacity'
/usr/bin/ld: ../../CMakeFiles/mbedtls_test.dir/tests/src/psa_exercise_key.c.o: in function `exercise_key_derivation_key':
psa_exercise_key.c:(.text+0x1f50): undefined reference to `psa_key_derivation_output_bytes'
/usr/bin/ld: psa_exercise_key.c:(.text+0x1f86): undefined reference to `psa_key_derivation_abort'
/usr/bin/ld: ../../CMakeFiles/mbedtls_test.dir/tests/src/psa_exercise_key.c.o: in function `mbedtls_test_psa_key_agreement_with_self':
psa_exercise_key.c:(.text+0x2048): undefined reference to `psa_get_key_attributes'
/usr/bin/ld: psa_exercise_key.c:(.text+0x2161): undefined reference to `psa_export_public_key'
/usr/bin/ld: psa_exercise_key.c:(.text+0x21a7): undefined reference to `psa_key_derivation_key_agreement'
/usr/bin/ld: psa_exercise_key.c:(.text+0x21bc): undefined reference to `psa_reset_key_attributes'
/usr/bin/ld: ../../CMakeFiles/mbedtls_test.dir/tests/src/psa_exercise_key.c.o: in function `mbedtls_test_psa_raw_key_agreement_with_self':
psa_exercise_key.c:(.text+0x2279): undefined reference to `psa_get_key_attributes'
/usr/bin/ld: psa_exercise_key.c:(.text+0x23d1): undefined reference to `psa_export_public_key'
/usr/bin/ld: psa_exercise_key.c:(.text+0x2436): undefined reference to `psa_raw_key_agreement'
/usr/bin/ld: psa_exercise_key.c:(.text+0x24d5): undefined reference to `psa_reset_key_attributes'
/usr/bin/ld: ../../CMakeFiles/mbedtls_test.dir/tests/src/psa_exercise_key.c.o: in function `exercise_key_agreement_key':
psa_exercise_key.c:(.text+0x260d): undefined reference to `psa_key_derivation_setup'
/usr/bin/ld: psa_exercise_key.c:(.text+0x2673): undefined reference to `psa_key_derivation_input_bytes'
/usr/bin/ld: psa_exercise_key.c:(.text+0x26ca): undefined reference to `psa_key_derivation_input_bytes'
/usr/bin/ld: psa_exercise_key.c:(.text+0x2761): undefined reference to `psa_get_key_attributes'
/usr/bin/ld: psa_exercise_key.c:(.text+0x29fe): undefined reference to `psa_key_derivation_input_bytes'
/usr/bin/ld: psa_exercise_key.c:(.text+0x2a65): undefined reference to `psa_key_derivation_input_bytes'
/usr/bin/ld: psa_exercise_key.c:(.text+0x2aab): undefined reference to `psa_key_derivation_output_bytes'
/usr/bin/ld: psa_exercise_key.c:(.text+0x2ae1): undefined reference to `psa_key_derivation_abort'
/usr/bin/ld: ../../CMakeFiles/mbedtls_test.dir/tests/src/psa_exercise_key.c.o: in function `exercise_export_key':
psa_exercise_key.c:(.text+0x34a1): undefined reference to `psa_get_key_attributes'
/usr/bin/ld: psa_exercise_key.c:(.text+0x36fb): undefined reference to `psa_export_key'
/usr/bin/ld: psa_exercise_key.c:(.text+0x3746): undefined reference to `psa_export_key'
/usr/bin/ld: psa_exercise_key.c:(.text+0x37b8): undefined reference to `psa_reset_key_attributes'
/usr/bin/ld: ../../CMakeFiles/mbedtls_test.dir/tests/src/psa_exercise_key.c.o: in function `exercise_export_public_key':
psa_exercise_key.c:(.text+0x3863): undefined reference to `psa_get_key_attributes'
/usr/bin/ld: psa_exercise_key.c:(.text+0x3ab7): undefined reference to `psa_export_public_key'
/usr/bin/ld: psa_exercise_key.c:(.text+0x3bd3): undefined reference to `psa_export_public_key'
/usr/bin/ld: psa_exercise_key.c:(.text+0x3c37): undefined reference to `psa_reset_key_attributes'
collect2: error: ld returned 1 exit status
make[2]: *** [programs/aes/CMakeFiles/crypt_and_hash.dir/build.make:121: programs/aes/crypt_and_hash] Error 1
make[1]: *** [CMakeFiles/Makefile2:970: programs/aes/CMakeFiles/crypt_and_hash.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
```

Should this maybe be in the toplevel `CMakeLists.txt`?

Reply:
Works for me now.

Since this wasn't trivial after all, this should be tested in `all.sh`. In any case we will need a test in `all.sh` if we start documenting it as supported functionality. We can add it to `component_test_cmake_out_of_source`. Preferably: edit `mbedtls_config.h` in the source tree, pass an alternative `MBEDTLS_CONFIG_FILE` in an out-of-tree build, and check that the configuration in the cmake build tree is affected by the `-DMBEDTLS_CONFIG_FILE` and not by the source tree. We should also test `MBEDTLS_USER_CONFIG_FILE` (I guess it can be done in the same build?).

Reply:
Is it useful to keep the number of components down? I was working on a separate component, but can move it into `test_cmake_out_of_source` if that's better.

Reply:
Ideally everything would be in separate components: it makes it easier to test stuff independently, and it makes reports from CI runs easier to understand.

But we have a persisting annoyance with the CI that there is a high cost to each component, which drives us towards unifying components.

On the gripping hand, we have a proposal for [making the CI group components](https://github.com/Mbed-TLS/mbedtls-test/issues/69), which would automatically resolve the performance issue with small components.